pkg/cloud/services/ec2/instances.go

```diff
@@ -910,6 +912,16 @@ func (s *Service) getInstanceAddresses(instance *ec2.Instance) []clusterv1.Machi
 			addresses = append(addresses, publicDNSAddress, publicIPAddress)
 		}
 	}
+
+	// Add secondary private DNS Name only if DHCP Option Set has domain name set
+	if domainName := s.GetDHCPOptionSetDomainName(s.EC2Client, instance.VpcId); domainName != nil {
+		additionalPrivateDNSAddress := clusterv1.MachineAddress{
+			Type:    clusterv1.MachineInternalDNS,
+			Address: fmt.Sprintf("%s.%s", strings.Split(privateDNSAddress.Address, ".")[0], *domainName),
+		}
+		addresses = append(addresses, additionalPrivateDNSAddress)
```

Comment:
I think this should be moved inside the first loop and use `eni.PrivateDnsName` instead of `privateDNSAddress.Address`